frame/contracts/src/exec.rs

```diff
@@ -735,19 +715,25 @@ where
 	type T = T;
 
 	fn get_storage(&self, key: &StorageKey) -> Option<Vec<u8>> {
-		self.ctx.overlay.get_storage(&self.ctx.self_account, self.ctx.self_trie_id.as_ref(), key)
+		match self.ctx.self_trie_id {
+			Some(ref trie_id) => storage::read_contract_storage(trie_id, key),
+			// TODO: My current understanding is that `self.ctx.self_trie_id` can be `None` only
+			// for the top-level account, i.e. EOA. As such, it cannot issue any reads to the
+			// storage, because it has no. Furthermore, EOA cannot have contract storage.
```

Comment:
Please do remove TODOs before merging. Your assumption is correct because there is no way to call `get_storage` from a top level `ExecutionContext` (and only these have `None`set). There is simply no `Dispatchable` that allows this. It always has to go through `Contracts::call` which then adds a nested `ExectionContext` with a self_trie_id.

That said, having the `self_trie_id` being an `Option` is annoying. Maybe we should split off a `RootExecutionContext` that does not have a `self_trie_id` and make it non-optional in `ExecutionContext`. That would save us a lot of those proofs and also catches problems at compile time.